### PR TITLE
[9.0.0] FIX for column auto size when the formulas result has changed width

### PR DESCRIPTION
**Pull request by Siemienik**
### Context
resolves: https://github.com/handsontable/handsontable/issues/8061

I added hook handler `afterFormulasValuesUpdate`, which the HyperFormula plugin emits after formulas calculation into the AutoColumnSize plugin. Then the handler maps event props into columns index and cache for further column calculation - it is the same logic as `onBeforeChange`. 

While the `afterFormulasValuesUpdate`is called only if HyperFormula is on, I considered turning off the `onBeforeChange` if HF is on. That reduces some calls on handlers. However, I won't do that because that additional condition is potentially risky and needs to follow another plugin state change. Additionally the `onBeforeChange` and `onAfterFormulasValuesUpdate` have quite simple logic.

May-27-2021 16-31-35

<sup>[skip changelog]</sup>

### How has this been tested?
I added three spec cases.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):
1. #8061 
2.
3.

### Affected project(s):
- [x] `handsontable`
- [ ] `@handsontable/angular`
- [ ] `@handsontable/react`
- [ ] `@handsontable/vue`

### Checklist:

- [x] I have reviewed the guidelines about [Contributing to Handsontable](https://github.com/handsontable/handsontable/blob/master/CONTRIBUTING.md) and I confirm that my code follows the code style of this project.
- [x] I have signed the [Contributor License Agreement](https://docs.google.com/forms/d/e/1FAIpQLScpMq4swMelvw3-onxC8Jl29m0fVp5hpf7d1yQVklqVjGjWGA/viewform?c=0&w=1)
- [ ] My change requires a change to the documentation.

### CODE REVIEW

**Review by aninde — CHANGES_REQUESTED** [review]
4 tests fail with this version so we should check it.

1.
```
Formulas general > Registering custom HF settings > should throw a warning when trying to register two named expressions under the same name (and register only the first one)
TypeError: Cannot read property 'col' of undefined
```

2.
```
Formulas general > Registering custom HF settings > should register the provided named expressions before creating the HF instance
TypeError: Cannot read property 'col' of undefined
```

3.
```
Formulas general > Engine-related hooks > afterNamedExpressionAdded > should run the `afterNamedExpressionAdded` hook when there were new named expressions added to the engine instance
TypeError: Cannot read property 'col' of undefined
```

4.
```
Comments > Hooks invoked after changing cell meta > should trigger `afterSetCellMeta` callback after editing comment by context menu
Expected spy afterSetCellMetaCallback to have been called with:
  [ 0, 0, 'comment', Object({ value: 'Edited comment' }), undefined, undefined ]
but it was never called.

```

**Comment by Siemienik** [comment]
@aninde Thank you for catching that. I fixed 1,2,3 cases by commit #daf1af7.

ad. failure 4. I checked it on my localhost, and this bug exists even if I disable the AutoColumnSize plugin (the place which I edited in this PR). 


https://jsfiddle.net/72r0Loef/1/ - example from this test, which stays on current develop.

Reproduce:
1. Open the context menu
2. Edit a comment
3. BUG: editing imminently being finished.

May-28-2021 12-33-00

File: .changelogs/8122.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "title": "Fixed column resizing when formulas results has changed width",
+  "type": "fixed",
+  "issue": 8122,
```
Comment (wojciechczerniak) [inline .changelogs/8122.json:4]: Should be issue number. Issue #8061 we're solving here is a regression on develop, was never released so from the client point of view having this in changelog will be confusing.

Comment (Siemienik) [reply .changelogs/8122.json:4]: ok, I remove it.

**Review by aninde — APPROVED** [review]
@Siemienik  thank you for the quick fixes and fast check. Now, this PR is perfect :)

File: src/plugins/autoColumnSize/autoColumnSize.js [hunk 2]
```diff
@@ -622,6 +624,15 @@ export class AutoColumnSize extends BasePlugin {
     privatePool.get(this).cachedColumnHeaders = this.hot.getColHeader();
   }
 
+  onAfterFormulasValuesUpdate(changes) {
+    const filteredChanges = arrayFilter(changes, change => change.address?.col !== undefined ?? false);
+    const changedColumns = arrayMap(filteredChanges, (change) => {
+      return this.hot.toPhysicalColumn(this.hot.propToCol(change.address.col));
+    });
```
Comment (jansiegel) [inline src/plugins/autoColumnSize/autoColumnSize.js:631]: I don't think `change.address.col` coming from HF can ever be a `string`, so this line's basically translating a physical column to a visual column and then back to a physical column.

I think this should work just as fine.
```suggestion
    const changedColumns = arrayMap(filteredChanges, change => change.address.col);
```

File: src/plugins/autoColumnSize/autoColumnSize.js [hunk 3]
```diff
@@ -622,6 +624,15 @@ export class AutoColumnSize extends BasePlugin {
     privatePool.get(this).cachedColumnHeaders = this.hot.getColHeader();
   }
 
+  onAfterFormulasValuesUpdate(changes) {
+    const filteredChanges = arrayFilter(changes, change => change.address?.col !== undefined ?? false);
```
Comment (jansiegel) [inline src/plugins/autoColumnSize/autoColumnSize.js:628]: Is `change.address?.col !== undefined ?? false` different from just `change.address?.col !== undefined`? (Or our helper's `isDefined(change.address?.col)`?)

Comment (Siemienik) [reply src/plugins/autoColumnSize/autoColumnSize.js:628]: `isDefined(change.address?.col)` will be the best choose

File: src/plugins/autoColumnSize/autoColumnSize.js (hunk 2 above)
Comment (Siemienik) [reply src/plugins/autoColumnSize/autoColumnSize.js]: Done in eb94c0a

File: src/plugins/autoColumnSize/autoColumnSize.js (hunk 3 above)
Comment (Siemienik) [reply src/plugins/autoColumnSize/autoColumnSize.js]: done in eb94c0a

**Review by jansiegel — APPROVED** [review]
LGTM 👌

**Review by jansiegel — APPROVED** [review]

File: src/plugins/autoColumnSize/autoColumnSize.js [hunk 4]
```diff
@@ -622,6 +625,13 @@ export class AutoColumnSize extends BasePlugin {
     privatePool.get(this).cachedColumnHeaders = this.hot.getColHeader();
   }
 
+  onAfterFormulasValuesUpdate(changes) {
```
Comment (budnix) [inline src/plugins/autoColumnSize/autoColumnSize.js:628]: Ohh, It seems that the linter didn't catch missing hook docs 🤔  Can you complete @Siemienik?

Comment (Siemienik) [reply src/plugins/autoColumnSize/autoColumnSize.js:634]: done in: 424ef10

**Review by jansiegel — APPROVED** [review]

**Review by budnix — APPROVED** [review]

**Pull request merged**